File: software-cracking-10weeks/code/week03/pe_entry_point.py
# -*- coding: utf-8 -*-
"""
03_pe_entry_point.py — Đọc ImageBase và AddressOfEntryPoint từ header PE.

Mục tiêu: minh hoạ cách tính entry point  EntryPoint = ImageBase + AddressOfEntryPoint
          mà không cần mở debugger.
Đầu vào : đường dẫn một file PE (chỉ ĐỌC).
Đầu ra : ImageBase, AddressOfEntryPoint, EntryPoint(VA), số section.
An toàn : chỉ đọc header, không thực thi, không ghi file.
"""
import struct
import sys


def read_pe_info(path: str) -> dict:
    with open(path, "rb") as f:
        data = f.read()

    # 1. Kiểm tra magic "MZ".
    if len(data) < 64 or data[:2] != b"MZ":
        raise ValueError("Không phải file PE (thiếu magic MZ).")

    # 2. e_lfanew nằm tại offset 0x3C, trỏ tới PE header.
    e_lfanew = struct.unpack_from("<I", data, 0x3C)[0]
    if data[e_lfanew:e_lfanew + 4] != b"PE\x00\x00":
        raise ValueError("Không tìm thấy chữ ký PE.")

    # 3. Optional header nằm ngay sau PE signature + COFF header (20 bytes).
    opt = e_lfanew + 4 + 20
    magic = struct.unpack_from("<H", data, opt)[0]  # 0x10B=PE32, 0x20B=PE32+

    # 4. Đọc ImageBase và AddressOfEntryPoint theo loại 32/64-bit.
    if magic == 0x10B:  # PE32
        image_base = struct.unpack_from("<I", data, opt + 28)[0]
        entry_rva = struct.unpack_from("<I", data, opt + 16)[0]
        num_sections = struct.unpack_from("<H", data, opt + 2)[0]
        arch = "PE32 (x86)"
    elif magic == 0x20B:  # PE32+
        image_base = struct.unpack_from("<Q", data, opt + 24)[0]
        entry_rva = struct.unpack_from("<I", data, opt + 16)[0]
        num_sections = struct.unpack_from("<H", data, opt + 2)[0]
        arch = "PE32+ (x64)"
    else:
        raise ValueError(f"Magic optional header lạ: 0x{magic:04X}.")

    return {
        "arch": arch,
        "image_base": image_base,
        "entry_rva": entry_rva,
        "entry_va": image_base + entry_rva,
        "num_sections": num_sections,
    }


def main() -> None:
    if len(sys.argv) < 2:
        print("Cách dùng: python pe_entry_point.py <file.exe>")
        sys.exit(1)

    try:
        info = read_pe_info(sys.argv[1])
    except (OSError, ValueError) as e:
        print(f"Lỗi: {e}")
        sys.exit(1)

    print(f"Kiến trúc             : {info['arch']}")
    print(f"ImageBase             : 0x{info['image_base']:08X}")
    print(f"AddressOfEntryPoint   : 0x{info['entry_rva']:08X} (RVA)")
    print(f"EntryPoint (VA)       : 0x{info['entry_va']:08X}")
    print(f"Số section            : {info['num_sections']}")
    print("\nCông thức: EntryPoint = ImageBase + AddressOfEntryPoint")


if __name__ == "__main__":
    main()

File: software-cracking-10weeks/code/week03/pe_entry_point.cpp
// 03_pe_entry_point.cpp — Đọc ImageBase + AddressOfEntryPoint từ PE (thật)
// Mục tiêu: tính EntryPoint = ImageBase + AddressOfEntryPoint, không cần debugger.
// An toàn: chỉ ĐỌC header, không thực thi.  g++ -std=c++17 -O2 pe_entry_point.cpp -o demo
#include <iostream>
#include <fstream>
#include <vector>
#include <cstdint>

uint32_t rd32(const std::vector<uint8_t>& d, size_t off) {
    return uint32_t(d[off]) | (uint32_t(d[off+1])<<8) | (uint32_t(d[off+2])<<16) | (uint32_t(d[off+3])<<24);
}
uint16_t rd16(const std::vector<uint8_t>& d, size_t off) { return uint16_t(d[off]) | (uint16_t(d[off+1])<<8); }

int main(int argc, char** argv) {
    if (argc < 2) { std::cout << "Cách dùng: ./demo <file.exe>\n"; return 1; }
    std::ifstream f(argv[1], std::ios::binary);
    std::vector<uint8_t> d((std::istreambuf_iterator<char>(f)), {});
    if (d.size() < 64 || !(d[0]=='M' && d[1]=='Z')) { std::cout << "Không phải PE.\n"; return 1; }

    uint32_t e_lfanew = rd32(d, 0x3C);
    if (d[e_lfanew]!='P' || d[e_lfanew+1]!='E') { std::cout << "Không tìm thấy chữ ký PE.\n"; return 1; }

    size_t opt = e_lfanew + 4 + 20;
    uint16_t magic = rd16(d, opt);
    if (magic == 0x10B) {                    // PE32
        uint32_t image_base = rd32(d, opt+28);
        uint32_t entry_rva  = rd32(d, opt+16);
        uint16_t nsect      = rd16(d, opt+2);
        std::cout << "Kiến trúc: PE32 (x86)\n";
        std::cout << "ImageBase: 0x" << std::hex << image_base << "\n";
        std::cout << "AddressOfEntryPoint: 0x" << entry_rva << " (RVA)\n";
        std::cout << "EntryPoint(VA): 0x" << (image_base + entry_rva) << std::dec << "\n";
        std::cout << "Số section: " << nsect << "\n";
    } else if (magic == 0x20B) {             // PE32+
        uint64_t image_base = rd32(d, opt+24);   // đọc 32-bit thấp (đơn giản hoá)
        uint32_t entry_rva  = rd32(d, opt+16);
        std::cout << "Kiến trúc: PE32+ (x64)\n";
        std::cout << "ImageBase: 0x" << std::hex << image_base << "\n";
        std::cout << "EntryPoint(VA): 0x" << (image_base + entry_rva) << std::dec << "\n";
    } else {
        std::cout << "Magic optional header lạ: 0x" << std::hex << magic << std::dec << "\n";
    }
    std::cout << "Công thức: EntryPoint = ImageBase + AddressOfEntryPoint\n";
    return 0;
}
